k-distribution/include/builtin/domains.k

```diff
@@ -590,6 +590,98 @@ module STRING-BUFFER
+
+  syntax Int ::= lengthBytes(Bytes) [function, hook(BYTES.length), smtlib(lengthBytes)]
+endmodule
+
+module BYTES-SYMBOLIC [symbolic]
+  imports BYTES-SYNTAX
+  imports INT
+  imports K-EQUAL
+  imports STRING
+  imports STRING-BUFFER
+
+  syntax Bytes ::= "nilBytes"
+                 | Int ":" Bytes
+
+  rule .Bytes => nilBytes
+
+  rule Bytes2Int(nilBytes, _, _) => 0
+  rule Bytes2Int(B : nilBytes, BE, Unsigned) => B
+  rule Bytes2Int(B0 : B1 : BS, BE, Unsigned) => Bytes2Int(((B0 <<Int 8) |Int B1) : BS, BE, Unsigned)
+  rule Bytes2Int(B0 : BS, BE, Signed) => signExtendBitRangeInt(Bytes2Int(B0 : BS, BE, Unsigned), 0, lengthBytes(B0 : BS) <<Int 3)
+  rule Bytes2Int(B0 : BS, LE, S) => Bytes2Int(reverseBytes(B0 : BS), BE, S)
+
+  syntax Bytes ::= Int2Bytes(Int, Bytes) [function, klabel(Int2BytesAux)]
+  rule Int2Bytes(LEN, I, BE) => padLeftBytes(Int2Bytes(bitRangeInt(I, 0, LEN <<Int 3), nilBytes), LEN, #if I <Int 0 #then 255 #else 0 #fi)
+  rule Int2Bytes(LEN, I, LE) => reverseBytes(Int2Bytes(LEN, I, BE))
+  rule Int2Bytes(0, BS) => BS
+  rule Int2Bytes(I, BS) => Int2Bytes(I >>Int 8, I &Int 255 : BS) requires I =/=Int 0
+
+  syntax String ::= Bytes2String(Bytes, StringBuffer) [function, klabel(Bytes2StringAux)]
+  rule Bytes2String(BS) => Bytes2String(BS, .StringBuffer)
+  rule Bytes2String(nilBytes, BUFFER) => StringBuffer2String(BUFFER)
+  rule Bytes2String(B : BS, BUFFER) => Bytes2String(BS, BUFFER +String chrChar(B))
+
+  rule String2Bytes(S) => ordChar(substrString(S, 0, 1)) : String2Bytes(substrString(S, 1, lengthString(S))) requires lengthString(S) >=Int 1
+  rule String2Bytes("") => nilBytes
+
+  rule substrBytes(BS, 0, 0) => nilBytes
+  rule substrBytes(B : BS, N, M) => substrBytes(BS, N -Int 1, M -Int 1) requires N >Int 0
+  rule substrBytes(B : BS, 0, M) => B : substrBytes(BS, 0, M -Int 1) requires M >Int 0
+
+  rule replaceAtBytes(BS, _, nilBytes) => BS
+  rule replaceAtBytes(B : BS, N, BS') => B : replaceAtBytes(BS, N -Int 1, BS') requires N >Int 0
+  rule replaceAtBytes(_ : BS, 0, B : BS') => B : replaceAtBytes(BS, 0, BS')
+
+  rule padRightBytes(BS, LEN, VAL) => reverseBytes(padLeftBytes(reverseBytes(BS), LEN, VAL))
+  rule padLeftBytes(BS, LEN, VAL) => BS requires lengthBytes(BS) >=Int LEN
+  rule padLeftBytes(BS, LEN, VAL) => padLeftBytes(VAL : BS, LEN, VAL) requires lengthBytes(BS) <Int LEN
+
+  syntax Bytes ::= reverseBytes(Bytes, Bytes) [function, klabel(reverseBytesAux)]
+  rule reverseBytes(BS) => reverseBytes(BS, nilBytes)
+  rule reverseBytes(nilBytes, BS) => BS
+  rule reverseBytes(B : BS, BS') => reverseBytes(BS, B : BS')
+
+  syntax Int ::= lengthBytes(Bytes, Int) [function, klabel(lengthBytesAux), smtlib(lengthBytesAux)]
+  rule lengthBytes(BS) => lengthBytes(BS, 0)
+  rule lengthBytes(nilBytes, SIZE) => SIZE
+  rule lengthBytes(B : BS, SIZE) => lengthBytes(BS, SIZE +Int 1)
+endmodule
+
+module BYTES
+  imports BYTES-SYNTAX
+  imports BYTES-SYMBOLIC
+  imports INT
```

Comment:
Is this `INT` import needed?

Reply:
yes, because we use log2Int in this module and the BYTES-SYMBOLIC import will not be present in the ocaml backend since it is a java-backend-only module.